When patching test class, don't run the original tests

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #103523

The problem is if you inherit from the test class you want
to transform, all its original test methods are still visible
and will get run again when you run the new test class.
There's not an easy way to "hide" them, so instead I just
don't inherit from the test class; instead, I "reconstruct"
the test class with the same bases and non-test methods.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

cc @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @aakhundov

ezyang committed:
When patching test class, don't run the original tests

The problem is if you inherit from the test class you want
to transform, all its original test methods are still visible
and will get run again when you run the new test class.
There's not an easy way to "hide" them, so instead I just
don't inherit from the test class; instead, I "reconstruct"
the test class with the same bases and non-test methods.

Signed-off-by: Edward Z. Yang <ezyang@meta.com>

[ghstack-poisoned]

clee2000 commented:
@pytorchbot rebase

Commit:
Update on "When patching test class, don't run the original tests"


The problem is if you inherit from the test class you want
to transform, all its original test methods are still visible
and will get run again when you run the new test class.
There's not an easy way to "hide" them, so instead I just
don't inherit from the test class; instead, I "reconstruct"
the test class with the same bases and non-test methods.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

Commit:
Update on "When patching test class, don't run the original tests"


The problem is if you inherit from the test class you want
to transform, all its original test methods are still visible
and will get run again when you run the new test class.
There's not an easy way to "hide" them, so instead I just
don't inherit from the test class; instead, I "reconstruct"
the test class with the same bases and non-test methods.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

Commit:
Update on "When patching test class, don't run the original tests"


The problem is if you inherit from the test class you want
to transform, all its original test methods are still visible
and will get run again when you run the new test class.
There's not an easy way to "hide" them, so instead I just
don't inherit from the test class; instead, I "reconstruct"
the test class with the same bases and non-test methods.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

albanD commented on test/dynamo/test_dynamic_shapes.py:
```diff
@@ -68,6 +68,7 @@ def make_dynamic_cls(cls):
 ]
 for test in tests:
     make_dynamic_cls(test)
+del globals()["test"]
```
del test should be enough here no?

albanD commented on test/dynamo/test_higher_order_ops.py:
```diff
@@ -1614,7 +1614,7 @@ def run(self, result=None):
         # testing on release branch.
         with config.patch(capture_func_transforms=True):
-            super().run(result)
+            super(type(self), self).run(result)
```
Why is this needed here? Isn't this never needed in 3.8+

albanD commented on torch/_dynamo/testing.py:
```diff
@@ -317,23 +317,24 @@ def _fn(*args, **kwargs):
             if xfail_prop is not None and hasattr(fn, xfail_prop):
                 new_fn = unittest.expectedFailure(new_fn)
             setattr(DummyTestClass, new_name, new_fn)
+        # NB: Doesn't handle slots correctly, but whatever
+        elif name not in DummyTestClass.__dict__:
```
```suggestion
        elif hasattr(DummyTestClass, name):
```

Would that solve the slot problem?

clee2000 replied:
Is there a preference here for has vs not has? I assumed we would want to copy over the things that don't exist

albanD commented on torch/_dynamo/testing.py:
```diff
@@ -317,23 +317,24 @@ def _fn(*args, **kwargs):
     DummyTestClass.__qualname__ = DummyTestClass.__name__
 
-    for name in dir(cls):
+    for name in cls.__dict__.keys():
```
Why not dir() anymore?

Commit:
Update on "When patching test class, don't run the original tests"


The problem is if you inherit from the test class you want
to transform, all its original test methods are still visible
and will get run again when you run the new test class.
There's not an easy way to "hide" them, so instead I just
don't inherit from the test class; instead, I "reconstruct"
the test class with the same bases and non-test methods.

Signed-off-by: Edward Z. Yang <ezyangmeta.com>

cc voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx ipiszy chenyang78 aakhundov

[ghstack-poisoned]

ezyang commented:
@clee2000 feel free to resubmit under your name so I can give you review juice
